### DISCUSSION TITLE
Working branches and the JJ "way"

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2425] [Q&A]
I've read the tutorial, and I understand that `jj` doesn't have a concept of a "working branch," but I'm wondering if I've missed something fundamental, because it seems like the basic git workflow of "make some changes, commit them, push them" is fairly high-friction.

Since `jj` doesn't have a concept of a "working branch", it appears that I need to explicitly run `jj branch set <name> -r @` every time I want to add commits to a branch. Is that correct? Is there a reason for that, other than encouraging users to rewrite commits rather than add new commits?

It may be that I just haven't adapted to the `jj` "way" (I only installed it yesterday), but here are the things I miss about having a working branch:

* Having a documented intention for the current thing I'm working on: `jj`'s `new` command does this for individual commits, but that seems too granular for me; I don't always know the nature of a small change I'm about to make before I make it. I do, however, always have a high-level plan, generally captured as a feature name and/or issue-tracker number.
* Easily and frequently pushing commits: I like that JJ captures the history of the working state in actual commits, but I tend to want my commits pushed to a remote as soon as possible so that I don't ever have to worry about forgetting to commit and losing work due to something bad happening to my local copy. Hard drives are probably more reliable now than they used to be, but I've gotten into the habit of always treating local device storage as purely ephemeral, and the online version of my data as the source of truth; and I've found this to be a pretty nice state of affairs even when nothing goes wrong. With JJ, `jj git push` doesn't work without either manually specifying a commit and a remote destination or first using `branch set ... -r @`. 

The second of these points is much more important to me than the first. Should I just get used to setting up `rsync`, Dropbox, or similar for my data-synchronization, and stop using commit & push for this purpose? Is `jj git push -c @` something I should be doing more frequently?

**Comment by BatmanAoD** [comment]
...actually, it seems even `branch set -r @` doesn't actually update the branch in the way I'd expect, but in fact causes pushing not to work if the working copy doesn't have a description, _even if the working copy is clean_. This seems...weird to me. In fact, it got me into a situation where I couldn't make the branch pushable without `--allow-backward`:

```
❯ : jj log
@  srwmpmxp batmanaod@gmail.com 2023-10-24 16:45:44.000 -06:00 my_branch_name* 215fe62f
│  (empty) (no description set)
◉  lzvvuyuz batmanaod@gmail.com 2023-10-24 16:45:44.000 -06:00 HEAD@git fe74a32b
│  some commit message
◉  wurtnqxy batmanaod@gmail.com 2023-10-24 16:30:30.000 -06:00 my_branch_name@origin 80e31921
│  another commit message

...

❯ : jj branch set my_branch_name -r @
Nothing changed.

❯ : jj git push
Error: Won't push commit 215fe62fd454 since it has no description

❯ : jj diff

❯ : jj diff my_branch_name@origin

❯ : jj status
The working copy is clean
Working copy : srwmpmxp 215fe62f my_branch_name* | (empty) (no description set)
Parent commit: lzvvuyuz fe74a32b some commit message

❯ : jj branch set my_branch_name -r HEAD@git
Error: Refusing to move branch backwards or sideways.
Hint: Use --allow-backwards to allow it.

❯ : jj branch set my_branch_name -r HEAD@git --allow-backwards

❯ : jj git push
Branch changes to push to origin:
  Move branch my_branch_name from 80e31921c6ad to fe74a32bcc9d
```

**Reply by ilyagr** [reply]
A few quick thoughts:

- If the working copy is empty, you should use ` jj branch set my_branch_name -r @-` (note the minus at the end).
- Using `--allow-backwards` is fine, as long as it's intentional; I do it all the time.
- Some people (like me) never use `jj git push -c`, others use it all the time. This does mean I use `jj branch set` relatively often (but not *that* often, see below). There are some proposals to make it less tedious.

One possibility is that you are used to having a long chain of commits where they are arranged in the order you made the changes. So, the 13th commit might be a fixup to the second commit in the chain. This would require a lot of `jj branch set`-s is indeed "not the jj way"; the jj way would be to reorder and squash commits where necessary.

There are a few workflows that people tend to use. https://martinvonz.github.io/jj/latest/FAQ/ and https://martinvonz.github.io/jj/v0.10.0/github/ might (or might not) be helpful, they describe some of them.

I generally create one branch per PR, after a `jj new`. After the branch already exists, I `jj new` to the commit I want to work on (often the branch tip) and then `jj amend && jj git push --all` often. No need for a `jj branch set`, unless I decide that the PR will be easier to review if it's split into several commits. (Then, I'd do one `jj branch set` per commit that will be in the PR)
 
I rely on `jj op log` and `jj obslog` to be able to recover previous states of the commits.

**Reply by BatmanAoD** [reply]
I appreciate the links to the FAQ and the GitHub/GitLab documentation; thanks.

**Reply by BatmanAoD** [reply]
For what it's worth, the author of this blog post also prefers long commit chains: https://v5.chriskrycho.com/journal/jj-init/

**Comment by martinvonz** [comment]
Thanks for taking the time to write feedback!

> Since `jj` doesn't have a concept of a "working branch", it appears that I need to explicitly run `jj branch set <name> -r @` every time I want to add commits to a branch. Is that correct? Is there a reason for that, other than encouraging users to rewrite commits rather than add new commits?

It's actually mostly because we haven't seen much need for it, so we're keeping the data model simpler until we're convinced it's needed. If there's such a thing as a "current branch" in the model, then you need to figure out when to set and when to clear that. Git sets the current branch when you pass the branch name to `git checkout/switch`, but not if you pass the commit id it points to. We would presumably copy that behavior if we added support for a current branch. We would also have to decide what should happen if you have the same current branch in multiple [workspaces](https://martinvonz.github.io/jj/prerelease/working-copy/#workspaces). There are probably a few other things I'm forgetting now.

> * Having a documented intention for the current thing I'm working on: `jj`'s `new` command does this for individual commits, but that seems too granular for me; I don't always know the nature of a small change I'm about to make before I make it. I do, however, always have a high-level plan, generally captured as a feature name and/or issue-tracker number.

Maybe we should support something like Mercurial's [topics extension](https://www.mercurial-scm.org/doc/evolution/tutorials/topic-tutorial.html), which lets you associate a single name with a commit. It lets you have the same topic name for multiple commits. Then you might do `jj new --topic foo main` to start working on topic "foo" off of branch "main", and any child commits created on top of a commit with a topic will inherit that topic (unless you pass `--topic <something else>`, of course).

There's also issue #2338 for adding a `jj commit --advance-branches` option. By the way, that should be really easy to implement (and I just added the "Good first issue" label to it), if you or someone else feels like it.

> * Easily and frequently pushing commits: [...]

That `--advance-branches` option might help if you often create small commits on top of the existing ones and want your branch to be moved forward for each commit so you can push it.

> The second of these points is much more important to me than the first. Should I just get used to setting up `rsync`, Dropbox, or similar for my data-synchronization, and stop using commit & push for this purpose? Is `jj git push -c @` something I should be doing more frequently?

I don't think you should have to use `rsync` or similar. We probably just need to make it simpler. `jj git push -c` will create a new branch every time you run it on a new commit, so you probably don't want that (you'd have to periodically delete the unwanted branches).

**Reply by necauqua** [reply]
Topics, as you've described them, sound pretty cool

Branches/tags point to a specific commit, and in jj are literally the same thing, and those topics point to multiple commits and propagate to their children once those are created

**Reply by arxanas** [reply]
Are there any specific advantages of topics over branches as a workflow? It seems like mainly that commits are implicitly added to topics (similar to committing to a branch head). I guess it also happens that a commit can belong to at most one topic (if I'm understanding the documentation correctly), which simplifies a lot of assumptions about e.g. which topics to push, but it seems like it might add friction when interacting with branches from external sources.

I do feel that topics might be a "better" abstraction than branches, in the sense that everyone considers branches to logically contain a set of commits, even though it's only a pointer to a single commit, so, with branches, you have to do a bunch of indirect algebra to calculate what commits are logically "part" of the branch (and rebase/merge that branch, etc.).

**Reply by joyously** [reply]
You are discussing branches from the Git and `jj` point of view. In the Breezy default, branches are folders. Breezy can do colocation like Git does, but it does not by default.
If topics are another way to tag or colocate versions, then it's simply more cognitive load for the user. I personally prefer less load on the human and more on the tool, like I get with folders as branches.

**Reply by arxanas** [reply]
> If topics are another way to tag or colocate versions

It's not clear to me from the topics documentation, but I think topics are a way to assign commits to arbitrary commit sets representing units of work. If I understand correctly, this concept doesn't exist as a first-class concept in Git or Breezy (or stock Mercurial or Monotone): sets of commits to be reviewed, etc., are implicitly calculated by computing the merge-base between two nodes in a commit graph and selecting one side. Possibly Pijul takes a more similar approach via some notion of patch sets.

To me topics system seems marginally better from a conceptual perspective (setting aside the fact that most people are familiar with Git-style branches and would have to learn a new concept), except that representing Git branches as topics for interop could be confusing if commits are meant to belong to no more than one topic.

> I personally prefer less load on the human and more on the tool, like I get with folders as branches.

I think we can already achieve this workflow with jj workspaces and simply not using branches? This should be much better-supported in jj than in Git. But I think it would be warranted to add a way to operate on remote commits directly (say something like `jj new https://my.remote/repo:abc123`), rather than having to fetch them/deal with remotes as a separate step.

**Reply by joyously** [reply]
>  Possibly Pijul takes a more similar approach via some notion of patch sets.

Yes, in theory Pijul channels are patch sets that you can work with separately. But in practice, they are polluted with the dependencies, and the choice was made to make everything colocated since most people are familiar with that from Git.

> I think we can already achieve this workflow with jj workspaces 

I haven't looked at workspaces. Is it not colocated?

**Reply by BatmanAoD** [reply]
I agree with @arxanas that topics sound pretty good for what I want, but also that it's not clear what the advantage of "topics" is versus just having branch-heads automatically advance like the active branch does in Git.

> That `--advance-branches` option might help if you often create small commits on top of the existing ones and want your branch to be moved forward for each commit so you can push it.

That is indeed the workflow I prefer, and it sounds pretty close to what I want. I've left a couple of comments in that issue thread.

----

> I haven't looked at workspaces. Is it not colocated?

It is not. https://martinvonz.github.io/jj/prerelease/working-copy/#workspaces

**Reply by BatmanAoD** [reply]
Here's an idea: what if the revset language included a symbol (or a function with a short name, though I'd prefer a symbol) to mean either "the last created or checked-out branch" or, if topics are introduced, "the branch(es) associated with the current topic"? That would mean that a lot of existing commands would more easily support working with branches, without needing users to type out branch names manually so often.

Additionally, as a possible alternative (or addition) to #2338, and inspired by the discussion about a possible `branch update` command in #2555, what if `jj branch update` had a `-t, --topic` or `-l, --latest` flag with similar behavior, to automatically update the topic and/or latest branch to `@-` (or, if non-empty, `@`)? In fact it could even include `-p, --push` or maybe `-s, --sync` so that you would only need one command to push commits instead of two.

**Reply by hgezim** [reply]
I hope this isn't too off topic (pun not intended), but for me, I want `--advance-branches` because I don't trust `jj log`. Things just seem to shift around a lot and I can't follow what's happening. Doing the behaviour OP described is my insurance policy. I also try to keep a git to on the side for the same reason.

**Comment by joyously** [comment]
Scott Chacon's [GitButler](https://gitbutler.com/) is "a Git client for simultaneous branches on top of your existing workflow". He talked about it in [So You Think You Know Git - FOSDEM 2024](urlhttps://www.youtube.com/watch?v=aolI_Rz0ZqY).
Since he's a co-founder of GitHub, and he's making a tool to do simultaneous branch access, I think my assessment is correct that branches shouldn't be colocated. To me, this is the biggest problem with Git. For `jj` to go the way of "no current branch" is sort of okay, but you still can't look at two different branches at the same time, like you can with Breezy.

**Reply by thoughtpolice** [reply]
Note that (as it was mentioned by Martin) you can use `jj workspace` to make branches appear as folders, much like how Darcs/Breezy (from your description) do it. Something like `jj workspace add -r name-of-branch path/to/workspace` should be OK, and will do what you ask. You can then just flip between the two folders as desired. Note that `path/to/workspace` can be any directory path at all, it does not need to exist underneath the existing repo (so `add -r fix-bug ../foo-branch-fix-bug` is OK, as well as `/absolute/path/to/place` too.) The workspace feature is basically the same as `git worktree` (though I never used it before.) Workspaces add an extra layer on top of branches, though, but I don't think they add too much.

I used Darcs for a long time, so I'm actually somewhat fond of the approach. Actually, in one of my projects, I use this trick extensively to both help CI automation (run some tests under a version without switching to a different working copy, workspaces are perfect for that) and as a development strategy. Sometimes it's really useful on certain projects where you might have expensive/time-consuming compilation steps. You can spend more disk space to have multiple folders of different versions and run `make` in each of them, or whatever.

Also, in recent times most people who have wanted "Virtual Branches" the way GitButler does it have now adopted a workflow I came up with a few months ago &mdash; though I think it came to me fairly naturally &mdash; which is called "Simultaneous Edits" in [@steveklabnik's tutorial](https://steveklabnik.github.io/jujutsu-tutorial/advanced/simultaneous-edits.html) (but I call it "Mega Merge".) The basic idea is you just create a merge commit on top of all the branches you want to check out simultaneously. Resolve any conflicts if needed, and then simply `jj squash --from @ --into ...` your changes back into the appropriate spot. You can use `jj rebase` to add or remove parents from the merge as you wish to update the "mega" checkout. This requires a bit more care; and it's assumed your branches are well defined already. It's a bit more work to e.g. create a new branch out of the existing change, and then adding it to the merge.

I haven't used GitButler, though I assume my way is more involved and more tedious. However I think these features arise from the basic fundamental building blocks that are already available, so you can sort of create it yourself. So maybe another question is, how do we achieve reusability out of the fundamental building blocks we have?

I'm not really sure if we should think about presenting these features differently. Just wanted to put this out there.

**Reply by bnjmnt4n** [reply]
FYI, I also wrote up another overview/walkthrough of the merge workflow: https://ofcr.se/jujutsu-merge-workflow

**Reply by jabdinghoff** [reply]
is there any discussion on making this workflow more straightforward? i find this a really compelling GitButler feature

**Reply by jennings** [reply]
@jabdinghoff What do you find not-straightforward about it? Do you have ideas about how it could be improved?

**Comment by steveklabnik** [comment]
Just tossing this out here: as someone working on a project using `jj` exclusively lately, working like this:

1. 1 change = 1 PR
2. "rebase" aka edit that change in response to feedback

makes all of this a breeze. The branches never move, so I never need to change them, a `jj git push` will just always push my branches up.

Depending on upstream policy, you may not be able to do this, but it ends up being *very* nice.

Since I like git's staging area, in practice I work on an empty `@` on top of a branch, and when I'm happy with changes, squash that back into the parent, then push.

**Reply by BatmanAoD** [reply]
Indeed I _have_ started using `rebase` a lot more liberally since writing this question. But I still find value in creating new commits on branches under review, partly because we have a team guideline not to destructively modify branches other people are actively looking at, and partly because I do sometimes appreciate having separate commits on a branch under review even if I ultimately plan to squash them.

**Reply by jimmiefulton-ybor** [reply]
I've been experimenting with modifying commits under PR, seeing what happens when other people have also checked out the PR bookmark and what happens when history is re-written.  In the past, I've always felt that "Never re-write pushed history".  But after trying this out, I can see that this really should be "Never re-write history on branches other people might base new work off of".

If I edit a commit in a PR and just `jj git push`, it does a "sideways movement".  The commit tree is clean.  If someone checks out the PR bookmark for review, and then later does a `jj git fetch`, their bookmark is updated.

The consequences of this are that Stacked PRs are straight-up easy.  If you have a chain of PRs up, and someone wants you to make a change to the very first link in the chain, no problem.  Make your changes.  If you get conflicts higher up in the chain of PRs, go fix the conflicts.  When you push, ALL of your PRs are updated with sideways changes to the bookmarks.  Damn!

**Reply by ejc3** [reply]
At Meta the modify PR flow is used like you describe, but the previous version is a first class citizen in the UI, so reviewers see what changed from the last time they reviewed it. And it support flows like “accept with suggested changes” where the engineer makes subsequent changes, pushes the branch and then a post commit review is triggered to review the diff between “accepted changes” and the set of changes actually committed.

**Reply by BatmanAoD** [reply]
@ejc3 GitHub and GitLab both have similar features, but in my experience they can be a little flaky, and neither platform makes it all that obvious to see how to do that kind of comparison, let alone including a link to do so in the notification itself. It's too bad the mainstream publicly-available forges are not better suited for stacked PRs.

**Comment by Coobaha** [comment]
I use the stacked workflow with `jj` and `revup` daily in various projects.

The only requirement is to add `topic` and `relative` trailers into the description:
```
Change

Topic: B

JJ: Relative only is needed if one wants to stack B on top of A, it is optional
Relative: A
```

[`revup`](https://github.com/Skydio/revup) takes care of creating branches and managing PRs

Most of the work has a linear history and is done without branching. Sometimes, I have a couple of stacks, but usually, one.

After customising templates, I also have these topics visible in the logs.

```
@  o 94f7
│  (empty) (no description set)
◉  y topic_C → topic_A 9cd0 HEAD@git
│  Change 4
◉  s topic_A 29dd 
│  Change 3 in Topic A
◉  n topic_B ec14
│  Change 2
◉  k topic_A 9313
│  Change 1
◉  w DRAFT another_topic bd01
│  This is not working
~
```

<details><summary><b>Config</b></summary>

```toml
[template-aliases]
topic = '''
  description.lines().map(|line|
      if(
          line.lower().starts_with("topic:"),
          line.substr(6, line.len()).remove_prefix(" ").remove_suffix(" "),
      )
  ).join("")
'''

relative_topic = '''
  description.lines().map(|line|
      if(
          line.lower().starts_with("relative:"),
          concat(
            "→ ",
            line.substr(9, line.len()).remove_prefix(" ").remove_suffix(" "),
          ),
      )
  ).join("")
'''

draft = '''
description.lines().map(|line|
      if(
          line.lower().starts_with("labels:") && line.lower().contains("draft"),
          "DRAFT"
      )
      ).join("")
'''

topic_label = 'separate(" ",  label("draft", draft), label("topic", topic), label("relative", relative_topic))'

[colors]
draft = { fg = "bright red", bold = false, underline = false }
topic = "bright yellow"
relative = "bright black"
"username" = { fg = "bright black", bold = false }
"timestamp" = { fg = "bright black", bold = false }

```
</details>

**Reply by BatmanAoD** [reply]
Splitting the "create branches and PRs" responsibility out to a separate tool makes a lot of sense to me! I hadn't heard of revup, so I will give that a try.

**Reply by BatmanAoD** [reply]
...ah, actually I can't use revup until it supports gitlab, at least for my current work. Unfortunate.

**Comment by nrbray** [comment]
This thread gave me the most clues about how to just get on and use jj, somehow, anyhow.  It has not been easy to find out what is the simplest way I can track my work in a remote git without breaking things.

To work on top of a branch, e.g. main, the simplest flow I have found is:

```sh
# after a jj git clone, or the previous jj git push, do:
# make changes
jj log # snapshots any changes made, causing jj to "stage" the changes - in jj it makes a detached head commit
jj commit -m "description of the changes made" # a conventional commit, but jj adds a fresh empty detached head commit on top
jj bookmark set --revision @- main # since we point to the fresh empty detached head commit on top, point the bookmark (branch name) to the prior not empty commit
jj git push -r @- # push the bookmark (git branch) that points to the prior not empty commit
# make more changes, and repeat
```

```jujutsu 
$ jj op log
@  e7f680213ffa nrb@retsnom 51 seconds ago, lasted 1 second
│  push bookmark main to git remote origin
│  args: jj git push -r @-
○  b040cd5d4c55 nrb@retsnom 2 minutes ago, lasted 102 milliseconds
│  point bookmark main to commit 65331304caf4939dab3856423ef58c9b0311998c
│  args: jj bookmark set --revision @- main
○  56b862ce9536 nrb@retsnom 4 minutes ago, lasted 100 milliseconds
│  commit 20fb932ec2534ee07e43a15d3bba65d3f5d65324
│  args: jj commit -m 'cast numeric, fix \'000\'s'
○  7c5a65cbef3a nrb@retsnom 7 minutes ago, lasted 15 milliseconds
│  snapshot working copy
│  args: jj log
○  df0f343ccfb0 nrb@retsnom 2 hours ago, lasted 950 milliseconds
│  push bookmark main to git remote origin
│  args: jj git push
```

**Reply by necauqua** [reply]
Any mutating jj command does snapshots first, so `jj log` is unnecesary (the whole point is that there's no staging, commands always see the repo state with @ being equal, in commit form, to the working copy dir), but other than that yeah that's how it goes.

One point is that you'd do series of small local commits and edits and rebases etc etc and then when you're going to push you update the bookmark before that - because you only need it for pushing, tugging the branch up after every commit (like you might or might not be doing) is unnecessary.
And for this update-bookmark-then-push there are shortcuts people do (in my case it looks like `jj tug && jj push` 🤷)

Another point - that other people do, not me - is that instead of `jj commit` at the end of work, they do `jj new -m 'commit message before actually doing something in that commit'`, and then either tug+push the @ or do one final `jj new` - because the commit subcommand is just a git-friendly shortcut for doing `jj desc && jj new`.
We actually got rid of `jj merge a b c` because it was just alias for `jj new a b c` - create new commit with parents a b and c

Topics will hopefully help as they're the "jj way" of having something auto-advancing with new commits that could be mapped to a git branch so you wont have to move it manually all the time.

**Reply by jimmiefulton-ybor** [reply]
What I've been doing is this:

jj git fetch
jj new main
Do some work
Describe it
If more work to do, jj new
Do some work
Describe it
Ready to check in....
jj b m main (if committing directly to main)
jj git push

Since I'm not using `jj commit`, I'm already on the commit that I want my bookmark to point to, so jj bookmark move <bookmark> just puts me on my current commit.  When I push, JJ automatically creates a new empty change, ready for me to do more work.

In this workflow, I just use "new" and "desc".  No commits.  Basically, I've gotten over the mental block that the current change is already a valid commit. I don't have to "new" or "commit" to create an empty change just to do a push.  It's perfectly fine to push the commit where HEAD is.

**Reply by BatmanAoD** [reply]
> When I push, JJ automatically creates a new empty change

That's something I didn't realize. It does print a warning, though, which suggests to me that maybe this isn't the intended workflow:

```
Warning: The working-copy commit in workspace 'default' became immutable, so a new commit has been created on top of it.
```

> jj b m main (if committing directly to main)

This is the part that I find confusing and tedious, though I didn't realize `m` worked as an alias for `move`/`set`, which does make this shorter. I push after _each commit_, so that at any time I can drop what I'm doing, come back, and I'll still have my work even if, in the meantime, my computer somehow gets fried. This hasn't happened to me recently, but it's happened to me before, and it has also happened to coworkers, and I just do not understand why, in this age of connectivity, anyone would want to have "local work" that isn't pushed somewhere.

So, for each commit, I want to do three things:

* describe the commit
* make a new one because I have more work to do, either now or in the future
* push what I have so far to a remote

Without `advance-branches` (which has some annoying problems that probably won't be resolved because it's an experimental feature), this requires at least three commands (`commit`, `bookmark`, `git`), and in your workflow it takes four (`new`, `describe`, `bookmark`, `git`), even if you don't personally do the `push` for every commit.

> In this workflow, I just use "new" and "desc". No commits.

`new` creates a commit.

**Reply by jimmiefulton-ybor** [reply]
Yeah, everyone has slightly different flow.  I actually _like_ that I have to mentally choose what I'm about to publish.  

In the git world, I've got a feature branch, and I push regularly.  The commits are ugly, but my work is saved.  It is very linear, and to actually care about what my commits look like is a lot of overhead I just can't be bothered with in many cases.

In the jj world, I find my mental model has completely shifted.  I look at my tree to review and reflect what everything looks like, and what I'm ready to publish and what I'm not ready to publish.  I can see the series of work, and make a mental note: "This is ready to be shared, PR'd, pushed to main", etc, and I move the bookmarks where I want them, and then push.

To your point, you want to make sure all your work is off of your machine.  For those long-lived branches, I have "feature" or "personal" branches that I use to push to.  Once again, as I'm wrapping up a section or session of work, I review the tree.  I've got some work that is ready to PR, so it gets a PR bookmark.  I've got some more work on top of that that I'm not ready to have reviewed, so it gets a personal bookmark.  jj git push to publish that view of the tree to the forge.

These tools really cater to a broad variety of workflows and personal preferences.  From my perspective, I was a little skeptical at first, because I had to forget what I was doing in git and rethink how I was doing DVCS.  But now that I've wrapped my head around it, it aligns much more with how working with git should be, even if I have to type an extra command.  I think the extra command is _necessary_, at least in my workflow, as part of a review step I should have been doing all along if my tool made that easy to do.  But I do see that some folks will have different flows, and having a bookmark follow automatically as you commit should be a viable way for those that prefer that approach.  I do wonder how that would work.  

As martinvonz has pointed out, what does it even mean to have an "active" bookmark in a world where bookmarks can be at a variety of places depending on your workflows.  I kinda like the idea of the `advance-branches`/`advance-bookmarks`, as this allows the command to be contextual to wherever you are, and can be added as an alias to allow everyone to customize their workflows as they see fit.

I'm aware that `new` creates a new commit.  I just choose to do `new` to start work, not to finish work.  I see the warning.  I really think that should be more informational than a warning.  You want the user to know that a new change has started, but it shouldn't be something scary. It's perfectly fine.

**Reply by necauqua** [reply]
@BatmanAoD for what you described, "push after _each_ commit", in my personal repos I do:
- doing some work
- `jj ci` to finalise that work into a commit at @- and give it a message
- `jj tug && jj push`, where tug is ~~my~~ (I didn't invent it, my own version only tugged `main`) alias for sneaky `jj b m` shown below and push is a dumb alias for `jj git push`

This does not look any more tedious than git to me, although I've been using jj exclusively for years now, maybe I forgot. And I get all the rebases/revsets/templates and other jj goodies.

And like, I don't _always_ push immediately, I might do some cleanups with `jj split -i` or `jj squash -i` or just create more neater commits etc etc, and only at the end I just do tug+push.

Well, as I kept and keep saying, topics - as I see them in my head anyway, and if/when they can be exported as git branches - will fix the auto-advancing issue anyway.

As, again, I've been saying that a lot too, bookmarks and tags are very similar - tags are named commit-ids (don't move at all unless you move them) and bookmarks are named change-ids (move with the changeid when rewriting commits, but don't otherwise, unless you move them).

So bookmarks semantically should not "auto-advance", we are just using bookmarks as a stop-gap to be able to push to git remotes, because git requires branches for pushes - and we will be using topics as a hack like that too, until the distant bright future where native jj repos are a thing, but at least they solve the auto-advancing.

**Reply by BatmanAoD** [reply]
> This does not look any more tedious than git to me

Well, it's three commands instead of two, and one of those is an alias you had to add. So, while admittedly a minor difference, it's strictly more onerous than git, especially since there isn't a built-in way to do it. And for what it's worth, I've previously used `git` with an alias that combined `add`, `commit`, and `push`, effectively doing what `svn commit` does. I recognize the value of distinguishing each step as a semantically-separate thing, but for most of my work I prefer the "I've made some changes, push them immediately" workflow.

> And like, I don't always push immediately...

I think that might just be the fundamental disconnect between myself and everyone happy with the "jj way." I _do always_ push immediately, even if I later rebase or otherwise modify commits. `jj` even makes it easy to push modified changes; it still ultimately does a force-push, but only if `jj` considers the change safe.

> topics... will fix the auto-advancing issue anyway.

Maybe, but it depends what the final design ends up being, and I have no idea how long it will take for topics to be available. In the meantime, it doesn't affect what I'm able to do _now_ with `jj`.

**Reply by necauqua** [reply]
Hm, actually I think you'd benefit greatly from a remote-backed commit store, since you push everything always immediately to have your work stored not only on your local machine I'm assuming.
In that case constantly pushing itself is tedious, not to mention the extra command, like you said.

With a remote-backed commit store (with a local layer ofc) you'd never have to push at all - well unless interacting with git repos. I guess that'd be basically a jj-aware server then, something along those lines

**Reply by BatmanAoD** [reply]
Yep, if you look at the last paragraph of my original post, I actually mentioned possibly doing this! I'll consider it again.

**Reply by jimmiebfulton** [reply]
Just to add other options to an interesting thread.  I'm really enjoying the flexibility and the config system.  I went through the publicly documented options, and among other things, I set the following aliases (more to come):

```toml
[aliases]
dds = ["diff", "--tool", "difft-s"]
ddi = ["diff", "--tool", "difft-i"]
di = ["diff", "--tool", "idea"]
sm = ["b", "s", "main"]
smb = ["b", "s", "main", "-B"]
gp = ["git", "push"]
gf = ["git", "fetch"]
gpn = ["git", "push", "--allow-new"]
```

I use code generation extensively, so I'm constantly (re)creating projects.  My workflow has been as follows:

```shell
jj git init --colocate
jj desc -m "initial commit"
jj sm # For "Set or Create the 'main' bookmark to Current Commit"
gh repo create --source . --public
jj gpn # For git push and allow new bookmarks
gh browse
```

With these aliases, 
* `jj sm` puts 'main' in the right place, even if it does not yet exist (I usually don't have empty commits at the HEAD)
* `jj smb` puts 'main' in an earlier point in time in the case where I've accidentally set it ahead of where I want
* `jj gf` fetches origin
* `jj gp` Push
* `jj gpn` Push new bookmarks
* `jj dds` Show a terminal diff in difftastic side-by-side
* `jj ddi` Show a terminal diff in difftastic inline
* `jj di` Show a diff in Intellj

To add:
* `jj ri` To resolve conflict in IntelliJ's very nice merge interface.

I'm sold.  Now I'm just contemplating on how I might be able to contribute to the project.  Docs, Rust is my primary language.  I've got an other open source project I'm in the midst of getting ready for publicizing (very powerful code generator), but I'm inclined to find some time for this, too.

**Comment by nrbray** [comment]
@necauqua Perfect, thank you. jj is fabulous and new to me.  I hope my beginner feedback helps.  From your comment I got 2 things:

1. I half understood that ```jj log``` didn't do anything for me here, but I think I was confused by the name commit for a command, and I think I am still troubled by the name of a command called ```commit``` in jj.  I might like it gone or maybe call it something else.  I see there is also a concept called *commit* in jj.  afaict The command ```commit``` modifies the description of a change and the concept *commit* associated with the change may be different in future. 

2. **jj tug** is 'obviously' the command I wanted instead of bookmark set, and was surprised in my hours with jj I had not come across it..  looks ;)  and remembers it is all still pre 1.0 release...

Like many, I am grateful for jujutsu appearing.  It feels like Déjà vu!  I lived through the godsend of git, after years of waiting for something like bzr, GNU Arch, Monotone, DARCS, Hg etc to go mainstream.

**Reply by necauqua** [reply]
Just to clarify, `jj tug` is not a command that exists, rather it's an alias that you could configure, tug just being a short name I chose for myself

You can have this in your `~/.config/jj/config.toml`
```toml
[aliases]
tug = ["bookmark", "move", "--from", "heads(::@- & bookmarks())", "--to", "@-"]
```
`jj bookmark move --from "heads(::@- & bookmarks())" --to @-` is a command that moves all bookmarks that are ancestors of `@-` to be at `@-`, and `jj tug` is a neat alias for it.

Reiterating a bit to help you with 1. - one of the main concepts of jj is that barring unusual circumstances commands work on the tree of commits, @ being just one of them.
Before most commands (where it makes sense) are run, changes to the working directory are automatically amended to that @ commit, so the command only sees the commits and doesn't have to interact with files at all.

Just working on commits and not having staging or index is both simple and very powerful - **all** commands can work on "remote" commits - I can have my @ being in one place, and then rebase some branch on top of some other branch and not have my work disrupted being one of many *many* examples of that.
Speaking of rebases - because of that they are done in memory aka waaay faster than how git does (or at least did?. I've been using jj exclusively in place of git for years already lol)

**Reply by simonmichael** [reply]
Thanks for `tug`. I attempted to refine it a little (not tugging to an empty change), and added an alias that also updates git's branch ref:

```toml
[aliases]

# tug - move the nearest bookmark to the nearest non-empty change at or before @
tug = ["bookmark", "move", "--from", "closest_bookmark(@)", "--to", "closest_nonempty(@)"]

# sync - move the nearest bookmark and current git branch to the nearest non-empty change at or before @
sync = ["util", "exec", "--", "bash", "-c", """
set -euo pipefail
jj tug
git switch `jj closestbookmark`
""", ""]

# closestbookmark - show the name of the nearest bookmark at or before @
closestbookmark = ["util", "exec", "--", "bash", "-c", """
jj log --no-graph -r 'closest_bookmark(@)' -T 'stringify(local_bookmarks.map(|b| b.name()).join("\n")).first_line()++"\n"'
""", ""]

[revset-aliases]

'closest_bookmark(to)' = 'heads(::to & bookmarks())'
'closest_nonempty(to)' = 'heads(::to ~ empty())'
```

**Comment by kajmagnus** [comment]
Here's a nice way to use jj with a Git style current-branch?:

Use `jj split -i` instead of `jj commit`.

Then, a bookmark pointing at `@` (the working-copy) stays at `@`, as if it was your currently active Git branch. No need for `jj branch set ...` or `bookmark move` :- )

And, as a bonus, reduces the likelihood that you accidentally commit the wrong files. I mean, `jj commit` commits all modified files, but with `split -i` you can have a quick look at what you're "committing". (I find the built-in patch picker https://github.com/arxanas/scm-record pretty nice for this: fast and keyboard shortcuts, nothing to install.) When using Git, I basically never do `git add .` or `git commit --all`, because of this risk.

Also, `jj reset` instead of `jj abandon` — keeps bookmarks pointing at `@` in place.

I'm pretty clueless about jj, still, I started wondering if such a `split -i` workflow could be simpler for [new people who are used to Git-style branches] to understand?  ... It's  an a bit confusing name though: "split" instead of "commit". Hmm.

The only "problem" is if there's two local-only bookmarks and you want to keep only _one_ pointing at `@`, but `split -i` "moves" both "forward", seems pretty rare to me? ... Oh, and you need to Git force-push.

Is this an a bit interesting tiny change of mindset? That a new feature is constantly represented by the same & stable change-id, and a "tail" keeps growing behind this change-id. Rather than adding new change-ids on top of each other?

**Reply by emesterhazy** [reply]
> Use `jj split -i` instead of `jj commit`.

Regarding this suggestion, we have an experimental feature that you can enable which will make `jj commit` and `jj new` automatically move branches forward.

* https://github.com/jj-vcs/jj/discussions/3549

**Reply by kajmagnus** [reply]
@emesterhazy Thanks! Trying it out now. Nice that it apparently works also with `commit --interactive` (so that I can still review everything being committed. Didn't know there is `commit -i` until just now)
